### PR TITLE
Refactor effect types

### PR DESCRIPTION
**Pull request by fwbrasil**
I'm working on documenting custom effects and the current naming using `Effect` and `ContextEffect` is a bit confusing. The `Effect` name doesn't communicate much meaning and I have to explain in the documentation that it's an abstract arrow.

This PR refactors the effect types with an abstract `Effect` and two sub-classes: `ArrowEffect` and `ContextEffect`.

### CODE REVIEW

File: kyo-prelude/shared/src/main/scala/kyo/kernel/ArrowEffect.scala [hunk 1]
```diff
@@ -0,0 +1,289 @@
+package kyo.kernel
+
+import internal.*
+import kyo.Tag
+import scala.annotation.tailrec
+import scala.util.control.NonFatal
+
+abstract class ArrowEffect[-I[_], +O[_]] extends Effect
```
Comment (fwbrasil) [inline kyo-prelude/shared/src/main/scala/kyo/kernel/ArrowEffect.scala:8]: The files are properly renamed in the git history but github doesn't render the rename. The individual commits should be easier to review.

File: kyo-zio/shared/src/main/scala/kyo/ZIOs.scala [hunk 2]
```diff
@@ -46,5 +46,5 @@ object ZIOs:
         }
     end run
 
-    sealed private[kyo] trait GetZIO extends Effect[ZIO[Any, Nothing, *], Id]
+    sealed private[kyo] trait GetZIO extends ArrowEffect[ZIO[Any, Nothing, *], Id]
```
Comment (hearnadam) [inline kyo-zio/shared/src/main/scala/kyo/ZIOs.scala:49]: Initially I was thinking this was a problem, but actually the context effects will be shifted into `Env` when we get line 20 working.

File: kyo-prelude/shared/src/main/scala/kyo/kernel/ArrowEffect.scala (hunk 1 above)
Comment (hearnadam) [reply kyo-prelude/shared/src/main/scala/kyo/kernel/ArrowEffect.scala:9]: I think your rebase may have broke this.

**Review by hearnadam — COMMENTED** [review]
Naming seems good to me. I'll check the diff closely once you fix the commits.

File: kyo-prelude/shared/src/main/scala/kyo/kernel/ArrowEffect.scala (hunk 1 above)
Comment (fwbrasil) [reply kyo-prelude/shared/src/main/scala/kyo/kernel/ArrowEffect.scala:9]: sorry, fixed now!

**Review by hearnadam — APPROVED** [review]

**Pull request merged**